### PR TITLE
Allow schedulers to step after validation

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

Fixes https://github.com/PyTorchLightning/pytorch-lightning/issues/5943
Allows schedulers to step after each time validation is performed. Beneficial when working with large datasets with `val_check_interval<1` and using the `ReduceLROnPlateau` scheduler . 

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/training_loop.py [hunk 1]
```diff
@@ -519,6 +520,10 @@ def run_training_epoch(self):
                 # reset stage to train
                 self.trainer._running_stage = RunningStage.TRAINING
 
+                # update val epoch schedulers
+                self.trainer.optimizer_connector.update_learning_rates(interval='val')
+                self.trainer.current_val_epoch += 1
```
Comment (carmocca) [inline pytorch_lightning/trainer/training_loop.py:525]: This logic should live inside run_evaluation

File: tests/trainer/optimization/test_optimizers.py [hunk 2]
```diff
@@ -467,3 +467,50 @@ def configure_optimizers(self):
...
+        val_check_interval=0.5
+    )
+    trainer.fit(model)
+    assert trainer.state == TrainerState.FINISHED, f"Training failed with {trainer.state}"
+
+    # check scheduler was configured correctly
+    sch = trainer.lr_schedulers
+    assert sch[0]['interval'] == interval, 'Learning rate scheduler not configured correctly'
+    import pdb
+    pdb.set_trace()
```
Comment (tchaton) [inline tests/trainer/optimization/test_optimizers.py:512]:
```suggestion
```

File: tests/trainer/optimization/test_optimizers.py [hunk 3]
```diff
@@ -467,3 +467,50 @@ def configure_optimizers(self):
...
+    
+    # class IntervalOptimizerModel(BoringModel):
+
+    #     def configure_optimizers(self):
+    #         opt = torch.optim.SGD(self.layer.parameters(), lr=init_lr)
+    #         lr_scheduler = torch.optim.lr_scheduler.StepLR(opt, 1, gamma=0.1)
+    #         return [opt], [{'scheduler': lr_scheduler, 'interval': interval}]
+
+    # model = IntervalOptimizerModel()
+
```
Comment (tchaton) [inline tests/trainer/optimization/test_optimizers.py:497]:
```suggestion
```

File: pytorch_lightning/trainer/training_loop.py [hunk 4]
```diff
@@ -69,6 +69,7 @@ def on_trainer_init(
         self.trainer.batch_idx = 0
         self.trainer.num_training_batches = 0
         self.trainer.train_dataloader = None
+        self.trainer.current_val_epoch = 0
```
Comment (tchaton) [inline pytorch_lightning/trainer/training_loop.py:72]: This should belong or set by `evaluation_loop`.

File: pytorch_lightning/trainer/trainer.py [hunk 5]
```diff
@@ -712,7 +712,10 @@ def run_evaluation(self, on_epoch=False):
         # update epoch-level lr_schedulers
         if on_epoch:
             self.optimizer_connector.update_learning_rates(interval='epoch')
-
+        elif not on_epoch and not self.sanity_checking:
+            self.optimizer_connector.update_learning_rates(interval='val')
+            self.current_val_epoch += 1
```
Comment (tchaton) [inline pytorch_lightning/trainer/trainer.py:717]: The increment should append inside update_learning_rates

File: pytorch_lightning/trainer/connectors/optimizer_connector.py [hunk 6]
```diff
@@ -36,8 +36,11 @@ def update_learning_rates(self, interval: str, monitor_metrics=None):
             return
 
         for scheduler_idx, lr_scheduler in enumerate(self.trainer.lr_schedulers):
-            current_idx = self.trainer.batch_idx if interval == 'step' else self.trainer.current_epoch
-            current_idx += 1  # account for both batch and epoch starts from 0
+            current_idx = {'step': self.trainer.batch_idx,
+                           'epoch': self.trainer.current_epoch,
+                           'val': self.trainer.current_val_epoch}[interval]
```
Comment (tchaton) [inline pytorch_lightning/trainer/connectors/optimizer_connector.py:41]: current_val_epoch is ambiguous as `val_check_interval<1`. `current_num_validation` or similar might be better.

File: pytorch_lightning/trainer/trainer.py [hunk 7]
```diff
@@ -712,7 +712,10 @@ def run_evaluation(self, on_epoch=False):
         # update epoch-level lr_schedulers
         if on_epoch:
             self.optimizer_connector.update_learning_rates(interval='epoch')
-
+        elif not on_epoch and not self.sanity_checking:
```
Comment (tchaton) [inline pytorch_lightning/trainer/trainer.py:715]: `interval='val'` won't be called `on_epoch=True`. I would assume it to be applied after each validation call.

**Review by tchaton — COMMENTED** [review]
Overall LGTM, small nits.

**Comment by rubencart** [comment]
Is there an update on this? Has it been merged already in another PR? Would be useful to me.
